### PR TITLE
Major version change for kubernetes task

### PR DESCRIPTION
**Pull request by Anumita**
Drop down to select azure kubernetes cluster
Remove json input from registry inside create secret section. Just have registry name.
Remove output variable section from kubernetes task.
Not download kubernets cli with every task  execution instead use the kubernetes utility in Common.

### CODE REVIEW

**Comment by bryanmacfarlane** [comment]
When you create a new major version, you need to create a SxS folder named <taskName>V2.

**Comment by bryanmacfarlane** [comment]
I see this is going from (0.) to (1.).  That means you just just rename the folder to KubernetesV1 as part of this change.  But, if you need to patch V0, you will need to create a KubernetesV0 folder when needed.  Your call on whether you just want to do it now.

Finally, new major versions need to start in preview (preview: true in task.json) and you need a marketing message on what the new major version brings (the UI will advertise, want to try new preview?  It's offers xxxx).

File: Tasks/Common/docker-common/registryauthenticationprovider/acrauthenticationtokenprovider.ts [hunk 1]
```diff
@@ -9,19 +9,28 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
...
+    private endpointName: string;
+
     constructor(endpointName?: string, registerNameValue?: string) {
         super();
 
         if(endpointName && registerNameValue) {
-            var obj = JSON.parse(registerNameValue);
+          try
+          {
+            var obj = JSON.parse(registerNameValue);
```
Comment (Anumita) [inline Tasks/Common/docker-common/registryauthenticationprovider/acrauthenticationtokenprovider.ts]: Add a debug log here too.

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        });
+    }
+
+    private async initialize(): Promise<void> {
+        if(!this.kubectlPath || !fs.existsSync(this.kubectlPath))
+        {
+            return this.getKubectl().then((kubectlpath)=> {
+                this.kubectlPath = kubectlpath; 
+                // prepend the tools path. instructs the agent to prepend for future tasks
+                if(!process.env['PATH'].startsWith(path.dirname(this.kubectlPath))) {
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: Ignore cases in starts with.

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        var kubeconfig = await this.getKubeConfig();
+         return this.initialize().then(() => {
+            this.kubeconfigFile = path.join(this.userDir, "config");
+            fs.writeFileSync(this.kubeconfigFile, kubeconfig);
+         });
+    }
+
+    // close kubernetes connection
+    public close(): void {
+        // all configuration are in agent temp directory. Hence automatically deleted.
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: Add a code to delete config file.

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            errlines.push(line);
+        });
+        return command.exec(options).fail(error => {
+            errlines.forEach(line => tl.error(line));
+            throw error;
+        });
+    }
+
+    // download kubernetes cluster config file from endpoint
+    private downloadKubeconfigFileFromEndpoint(kubernetesEndpoint: string) : void {
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: Delete this function.

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        this.kubeconfigFile = path.join(this.userDir, "config");
+        var kubeconfig = tl.getEndpointAuthorizationParameter(kubernetesEndpoint, 'kubeconfig', false);
+        fs.writeFileSync(this.kubeconfigFile, kubeconfig);
+    }
+
+    private getExecutableExtention(): string {
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: Delete this function

File: Tasks/KubernetesV1/task.json [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    ],
+    "inputs": [
+        {
+            "name": "connectionType",
+            "type": "pickList",
+            "label": "Connection type",
+            "defaultValue": "Azure Resource Manager",
+            "required": true,
+            "options": {
+                "Azure Resource Manager": "Azure Resource Manager",
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: Add a None here. Get the exact name from Atul/

File: Tasks/KubernetesV1/task.json [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                "Kubernetes Service Connection": "Kubernetes Service Connection"
+            },
+            "helpMarkDown": "Select a connection type."
+        },
+        {
+            "name": "kubernetesServiceEndpoint",
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: Mark as required.

File: Tasks/KubernetesV1/task.json [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+            "helpMarkDown": "Working directory for the Kubectl command.",
+            "groupName": "advanced"
+        }       
+    ],
+    "dataSourceBindings": [
+        {
+            "target": "azureContainerRegistry",
+            "endpointId": "$(azureSubscriptionEndpoint2)",
+            "dataSourceName": "AzureRMContainerRegistries",
+            "resultTemplate": "{{{properties.loginServer}}}"
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: Lets show name of the azure registry in the ui and save the url as value.

File: .gitignore [hunk 9]
```diff
@@ -7,8 +7,7 @@ gulp-tsc-tmp-*
 
 [Tt]asks/**/*.js
 [Tt]ests/**/*.js
-!Tasks/AzureResourceGroupDeployment*/Tests/mock_node_modules/**/*.js
```
Comment (jikuma) [inline .gitignore]: Why was it required to make the change in this file?

File: Tasks/ANT/Strings/resources.resjson/en-US/resources.resjson [hunk 10]
```diff
@@ -1,45 +0,0 @@
-{
-  "loc.friendlyName": "Ant",
```
Comment (jikuma) [inline Tasks/ANT/Strings/resources.resjson/en-US/resources.resjson]: Why did we need to delete this file.

File: Tasks/KubernetesV1/Tests/TestShared.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+export let ContainerTypes = {
+    AzureContainerRegistry: "Azure Container Registry",
+    ContainerRegistry: "Container Registry"
+}
+
+export let ConnectionType = {
+    AzureResourceManager: "Azure Resource Manager",
+    KubernetesServiceConnection: "Kubernetes Service Connection",
+    NotApplicable: "Not Applicable"
```
Comment (Anumita) [inline Tasks/KubernetesV1/Tests/TestShared.ts]: None

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        else
+        {
+            return null;
+        }
+    }
+    
+    // get kubeconfig file path
+    private async getKubeConfig(): Promise<string> {
+        var connectionType = tl.getInput("connectionType", true);
+        if(connectionType != 'Not Applicable')
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: Remove Not applicable check and return null

File: Tasks/KubernetesV1/src/clusterconnection.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    private kubectlPath: string;
+    private kubeconfigFile: string;
+    private userDir: string;
+
+    constructor() {
+        this.kubectlPath = tl.which("kubectl", false);
+        this.userDir = utils.getNewUserDirPath();
+    }
+
+    private getClusterType(connectionType: string): any {
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusterconnection.ts]: loadClusterType

File: Tasks/KubernetesV1/src/clusters/armkubernetescluster.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    
+    tl.debug(tl.loc("KubernetesClusterResourceGroup", clusterName, resourceGroup));
+
+    var clusterInfo : AKSClusterAccessProfile = await aks.getAccessProfile(resourceGroup, clusterName);
+    var Base64 = require('js-base64').Base64;
+    return Base64.decode(clusterInfo.properties.kubeConfig);
+}
+
+export async function getKubeConfig(): Promise<string> {
+    var clusterName : string = tl.getInput("kubernetesCluster", true);
```
Comment (Anumita) [inline Tasks/KubernetesV1/src/clusters/armkubernetescluster.ts]: Move it below tl.getInput("azureResourceGroup"...)

File: Tasks/KubernetesV1/task.json [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+        "Release"
+    ],
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 1,
+        "Minor": 0,
+        "Patch": 0
+    },
+    "demands": [],
+    "releaseNotes": "What's new in Version 1.0: <br/>&nbsp;&nbsp; Use Azure service connection to select targeted AKS cluster.<br/>&nbsp;&nbsp;Output section has been removed as there already is an Output variables section.",
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: Reword "What's new in Version 1.0: <br/>&nbsp;&nbsp; Use Azure service connection to select targeted AKS cluster.<br/>&nbsp;&nbsp;Output section has been removed as there already is an Output variables section." to 


What's new in Version 1.0: <br/>&nbsp;&nbsp; Added new connection type input for easy selection of Azure AKS cluster.<br/>&nbsp;&nbsp;Replaced output variable input with output variables section that we had added in all tasks

""

Comment (Anumita) [reply Tasks/KubernetesV1/task.json]: Fix the formatting

File: Tasks/KubernetesV1/task.json [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+            "label": "Working directory",
+            "defaultValue": "$(System.DefaultWorkingDirectory)",
+            "helpMarkDown": "Working directory for the Kubectl command.",
+            "groupName": "advanced"
+        }       
+    ],
+    "dataSourceBindings": [
+        {
+            "target": "azureContainerRegistry",
+            "endpointId": "$(azureSubscriptionEndpoint2)",
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: If user has selected the AKS connection type and user selects docker registry in secrets, do we really need to ask for Azure container registry and azure service endpoint.

File: Tasks/KubernetesV1/task.json [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+        {
+            "target": "azureResourceGroup",
+            "endpointId": "$(azureSubscriptionEndpoint)",
+            "endpointUrl":"{{{endpoint.url}}}/subscriptions/{{{endpoint.subscriptionId}}}/providers/Microsoft.ContainerService/managedClusters?api-version=2017-08-31",
+            "resultSelector": "jsonpath:$.value[*]",
+            "resultTemplate": "{{{ #extractResource id resourcegroups}}}"
+        }
+    ],
+    "instanceNameFormat": "kubectl $(command)",
+    "outputVariables": [
```
Comment (Anumita) [inline Tasks/KubernetesV1/task.json]: bring back output format in the advanced section

**Comment by stephenmichaelf** [comment]
It's probably easier to rebase on latest master.

**Pull request closed without merging**